File: fredrik/day01.py
import re

from shared import utils

NUM_MAP = {
    "one": "1",
    "two": "2",
    "three": "3",
    "four": "4",
    "five": "5",
    "six": "6",
    "seven": "7",
    "eight": "8",
    "nine": "9",
}

REGEX_NUMS = r"\d"
REGEX_LETTERS_NUMS = f"{'|'.join(NUM_MAP.keys())}|{REGEX_NUMS}"


def calibration_value(line: str, regex: str) -> int:
    matches = re.finditer(f"(?=({regex}))", line)
    digits = [match.group(1) for match in matches]
    first, last = digits[0], digits[-1]

    first = NUM_MAP[first] if first in NUM_MAP else first
    last = NUM_MAP[last] if last in NUM_MAP else last

    return int(f"{first}{last}")


def part1(lines: list[str]) -> int:
    return sum(calibration_value(line=line, regex=REGEX_NUMS) for line in lines)


def part2(lines: list[str]) -> int:
    return sum(calibration_value(line=line, regex=REGEX_LETTERS_NUMS) for line in lines)


def main() -> None:
    lines = utils.read_input_to_string().splitlines()

    print(f"Part 1: {part1(lines=lines)}")
    print(f"Part 2: {part2(lines=lines)}")


if __name__ == "__main__":
    main()

File: kaugesaar/2023/day01/day01.go
package day01

import (
	"kaugesaar-aoc/solution"
	"kaugesaar-aoc/utils"
	"strings"
	"unicode"
)

// Solver for day 1 and its both parts
type Solver struct{}

func parser(part int) string {
	rows := utils.ReadFile("day1.txt")
	var numbers []int

	findFirst := func(s string) string {
		for i := 0; i < len(s); i++ {
			if unicode.IsDigit(rune(s[i])) {
				return string(s[i])
			}
		}
		return ""
	}

	findLast := func(s string) string {
		for i := len(s) - 1; i >= 0; i-- {
			if unicode.IsDigit(rune(s[i])) {
				return string(s[i])
			}
		}
		return ""
	}

	for _, row := range rows {
		if part == 2 {
			row = replacer(row)
		}

		first := findFirst(row)
		last := findLast(row)

		number := utils.ToInt(first + last)
		numbers = append(numbers, number)
	}

	return utils.ToStr(utils.SumArr(numbers))
}

func replacer(s string) string {
	replaceMap := map[string]string{
		"one":   "o1e",
		"two":   "t2o",
		"three": "t3e",
		"four":  "4",
		"five":  "5e",
		"six":   "6",
		"seven": "7n",
		"eight": "e8t",
		"nine":  "n9e",
	}

	for key, value := range replaceMap {
		s = strings.ReplaceAll(s, key, value)
	}

	return s
}

// Part1 the solution for part 1, day 1
func (s Solver) Part1() solution.Response {
	return solution.Response{
		Day:    1,
		Part:   1,
		Answer: parser(1),
	}
}

// Part2 the solution for part 2, day 1
func (s Solver) Part2() solution.Response {
	return solution.Response{
		Day:    1,
		Part:   2,
		Answer: parser(2),
	}
}
